File: python_test_crate/tests/02_interface.py
from typing import Any, Optional, List, Dict, Callable

class WithInterface:
    def __init__(self):
        self._state_stack = []
        self._state_context = {}
        self._return_value = None
        self.call_count = 0
        self._state = "Ready"
        self._enter()

    def _transition(self, target_state, exit_args = None, enter_args = None):
        if exit_args:
            self._exit(*exit_args)
        else:
            self._exit()
        self._state = target_state
        if enter_args:
            self._enter(*enter_args)
        else:
            self._enter()

    def _change_state(self, target_state):
        self._state = target_state

    def _dispatch_event(self, event, *args):
        handler_name = f"_s_{self._state}_{event}"
        handler = getattr(self, handler_name, None)
        if handler:
            return handler(*args)

    def _enter(self, *args):
        # No enter handlers
        pass

    def _exit(self, *args):
        # No exit handlers
        pass

    def greet(self, name: str) -> str:
        self._return_value = None
        self._dispatch_event("greet", name)
        return self._return_value

    def get_count(self) -> int:
        self._return_value = None
        self._dispatch_event("get_count")
        return self._return_value

    def _s_Ready_greet(self, name: str) -> str:
        self.call_count += 1
        self._return_value = f"Hello, {name}!"
        return

    def _s_Ready_get_count(self) -> int:
        self._return_value = self.call_count
        return


def main():
    print("=== Test 02: Interface Methods ===")
    s = WithInterface()

    # Test interface method with parameter and return
    result = s.greet("World")
    assert result == "Hello, World!", f"Expected 'Hello, World!', got '{result}'"
    print(f"greet('World') = {result}")

    # Test domain variable access through interface
    count = s.get_count()
    assert count == 1, f"Expected count=1, got {count}"
    print(f"get_count() = {count}")

    # Call again to verify state
    s.greet("Frame")
    count2 = s.get_count()
    assert count2 == 2, f"Expected count=2, got {count2}"
    print(f"After second call: get_count() = {count2}")

    print("PASS: Interface methods work correctly")

if __name__ == '__main__':
    main()

File: typescript_test_crate/tests/02_interface.ts
class WithInterface {
    private _state: string;
    private _state_stack: Array<any>;
    private _state_context: Record<string, any>;
    private _return_value: any;
    private call_count: number = 0;

    constructor() {
        this._state_stack = [];
        this._state_context = {  };
        this._return_value = null;
        this.call_count = 0;
        this._state = "Ready";
        this._enter();
    }

    private _transition(target_state: string, exit_args: any = null, enter_args: any = null) {
        if (exit_args) {
            this._exit(...exit_args);
        } else {
            this._exit();
        }
        this._state = target_state;
        if (enter_args) {
            this._enter(...enter_args);
        } else {
            this._enter();
        }
    }

    private _change_state(target_state: string) {
        this._state = target_state;
    }

    private _dispatch_event(event: string, ...args: any[]) {
        const handler_name = `_s_${this._state}_${event}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            return handler.apply(this, args);
        }
    }

    private _enter(...args: any[]) {
        // No enter handlers
    }

    private _exit(...args: any[]) {
        // No exit handlers
    }

    public greet(name: string): string {
        this._return_value = null
        this._dispatch_event("greet", name)
        return this._return_value
    }

    public get_count(): number {
        this._return_value = null
        this._dispatch_event("get_count")
        return this._return_value
    }

    private _s_Ready_get_count() {
        this._return_value = this.call_count;
        return this._return_value;;
    }

    private _s_Ready_greet(name: string) {
        this.call_count += 1;
        this._return_value = `Hello, ${name}!`;
        return this._return_value;;
    }
}


function main() {
    console.log("=== Test 02: Interface Methods ===");
    const s = new WithInterface();

    // Test interface method with parameter and return
    const result = s.greet("World");
    if (result !== "Hello, World!") {
        throw new Error(`Expected 'Hello, World!', got '${result}'`);
    }
    console.log(`greet('World') = ${result}`);

    // Test domain variable access through interface
    let count = s.get_count();
    if (count !== 1) {
        throw new Error(`Expected count=1, got ${count}`);
    }
    console.log(`get_count() = ${count}`);

    // Call again to verify state
    s.greet("Frame");
    const count2 = s.get_count();
    if (count2 !== 2) {
        throw new Error(`Expected count=2, got ${count2}`);
    }
    console.log(`After second call: get_count() = ${count2}`);

    console.log("PASS: Interface methods work correctly");
}

main();

